Adding yangdocs pyang plugin

aashaikh opened the pull request:
Requesting merge of yangdocs plugin,  associated libs, and web files.

aashaikh committed:
Adding yangdocs pyang plugin, associated utility scripts, and
html/css/js template files.

robshakir commented on plugins/templates/js/yangdoc.menu.js:
```diff
@@ -0,0 +1,87 @@
+});
+
+
+//Initialization of treeviews
```
Should we clean this comment out?

Additional question - have there been modifications to this file (and the menu CSS) that we need to mark with a different copyright?

robshakir commented on plugins/templates/yangdoc.html:
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE html>
```
Does this file need a copyright?

robshakir commented on plugins/util/doc_emitter.py:
```diff
@@ -0,0 +1,51 @@
+
+class DocEmitter:
+
+  def __init__ (self):
```
`def __init__(self)` - i.e., remove the space before the bracket :-)

robshakir commented on plugins/util/html_emitter.py:
```diff
@@ -0,0 +1,365 @@
+
+        # collect all of the identities that have base_id as
+        # their base
+        derived = { key:value for key,value in mod.identities.items() if value.attrs['base'] == base_id }
```
As mentioned before, I'll try and create a library here that finds all resolved identity names since this won't do multiple inheritance.

robshakir commented on plugins/util/html_emitter.py:
```diff
@@ -0,0 +1,365 @@
+    navs = []
+    navids = []
+    # create the documentation elements for each module
+    for module_name in self.moduledocs.keys():
```
FYI, you should just do `for module_name in self.moduledocs` in general. A dict has an iterator that has just the keys by default, if you use `.keys()` for this then it creates a list object of those keys, which generally performs worse.

Since there's lots of dict access here too, it can be cleaner to do `for module_name, module_definition in self.moduledocs.iteritems()` which gives you both the name and then a `module_definition` object, which is the value of the dict entry.

robshakir commented on plugins/util/html_emitter.py:
```diff
@@ -0,0 +1,365 @@
+
+  return nav
+
+def text_to_paragraphs (textblock):
```
Space between name and brackets.

robshakir commented on plugins/util/html_emitter.py:
```diff
@@ -0,0 +1,365 @@
+  paras = textblock.split("\n\n")
+  return paras
+
+def node_to_id (node):
```
Ditto on the space.

robshakir commented on plugins/util/html_helper_test.py:
```diff
@@ -0,0 +1,71 @@
+import html_helper
```
Copyright.

robshakir commented on plugins/util/html_helper_test.py:
```diff
@@ -0,0 +1,71 @@
+  tempus sed risus ac, lobortis efficitur lorem."
+
+  ht = html_helper.HTMLHelper()
+  print ht.h1(text)
```
You asked a while ago about `unittest`, remind me and we can talk through that.

robshakir commented on plugins/util/markdown_helper_test.py:
```diff
@@ -0,0 +1,34 @@
+import markdown_helper
```
Copyright again.

aashaikh committed:
Several fixes for yangdocs plugin files based on review feedback:
 - added / fixed missing copyright notices
 - fixed spacing issues
 - removed several uses of <dict>.keys()
 - reformatted statement, type, identity docs to list description first
 - added path prefix just above each statement name (issue #6 in oc-tools repo)
 - added utility function to separate last path element from prefix in yangpath lib

aashaikh commented:
@robshakir Updated files to address all of the comments.  PTAL.

latest docs can be seen [here](http://ops.openconfig.net/docs/)

robshakir reviewed (approved):
LGTM
